### PR TITLE
[SYCL][Fusion] Cache JIT compiled fused kernels

### PR DESCRIPTION
**Pull request by sommerlukas**
Avoid repeating the kernel fusion JIT compilation pipeline if the sequence of kernels to be fused is equivalent to a sequence of kernels fused before. 

Next to the sequence of kernels, the detected parameter identities & constants as well as the user-specified internalizations play a role for the equivalence of two sequences. This is all incorporated in the cache lookup. 

The JIT compilation caching (on/off) can be controlled through the configuration/environment variable `SYCL_ENABLE_FUSION_CACHING`, the default is on. 

Signed-off-by: Lukas Sommer <lukas.sommer@codeplay.com>

### CODE REVIEW

**Comment by sommerlukas** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1416

**Comment by sommerlukas** [comment]
Putting up the test for this PR is blocked by review & merge of https://github.com/intel/llvm-test-suite/pull/1416. 

There is draft PR [here](https://github.com/sommerlukas/llvm-test-suite/pull/2), based on another [draft PR](https://github.com/sommerlukas/llvm-test-suite/pull/1) with the tests for https://github.com/intel/llvm/pull/7831, which in turn is blocked by https://github.com/intel/llvm-test-suite/pull/1416.

File: sycl-fusion/jit-compiler/include/Hashing.h [hunk 1]
```diff
@@ -0,0 +1,98 @@
+//==---- Hashing.h - helper for hashes for JIT internal representations ----==//
```
Comment (victor-eds) [inline sycl-fusion/jit-compiler/include/Hashing.h:1]: Please, use `llvm/ADT/Hashing.h` mechanisms in this file.

File: sycl-fusion/jit-compiler/lib/JITContext.cpp [hunk 2]
```diff
@@ -33,3 +33,16 @@ SPIRVBinary &JITContext::emplaceSPIRVBinary(std::string Binary) {
...
+  ReadLockT ReadLock{CacheMutex};
+  if (Cache.count(Identifier)) {
+    return Cache.at(Identifier);
+  }
+  return {};
+}
+
+void JITContext::addCacheEntry(CacheKeyT &Identifier, SYCLKernelInfo &Kernel) {
+  WriteLockT WriteLock{CacheMutex};
+  Cache.emplace(std::make_pair(Identifier, Kernel));
```
Comment (victor-eds) [inline sycl-fusion/jit-compiler/lib/JITContext.cpp:47]:
```suggestion
  Cache.emplace(Identifier, Kernel);
```
Use `emplace` to build the pair

**Review by victor-eds — CHANGES_REQUESTED** [review]

**Review by victor-eds — APPROVED** [review]

File: sycl/source/detail/config.def [hunk 3]
```diff
@@ -39,3 +39,4 @@ CONFIG(SYCL_QUEUE_THREAD_POOL_SIZE, 4, __SYCL_QUEUE_THREAD_POOL_SIZE)
 CONFIG(SYCL_RT_WARNING_LEVEL, 4, __SYCL_RT_WARNING_LEVEL)
 CONFIG(SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE, 16, __SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE)
 CONFIG(ONEAPI_DEVICE_SELECTOR, 1024, __ONEAPI_DEVICE_SELECTOR)
+CONFIG(SYCL_ENABLE_FUSION_CACHING, 1, __SYCL_ENABLE_FUSION_CACHING)
```
Comment (steffenlarsen) [inline sycl/source/detail/config.def:42]: This new option should be documented in sycl/doc/EnvironmentVariables.md.

File: sycl/source/detail/config.hpp [hunk 4]
```diff
@@ -579,6 +579,34 @@ template <> class SYCLConfig<SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE> {
...
+public:
+  static bool get() {
+    constexpr bool DefaultValue = true;
+
+    const char *ValStr = getCachedValue();
+
+    if (!ValStr)
+      return DefaultValue;
+
+    return std::stoi(ValStr);
```
Comment (steffenlarsen) [inline sycl/source/detail/config.hpp:594]: It should also be specified in the aforementioned documentation of the option, but what is the expected behavior if I pass some non-integer value? `std::stoi` will throw, so if we want it to fall back to the default value in that case we would probably want to catch it.

File: sycl/source/detail/config.def (hunk 3 above)
Comment (sommerlukas) [reply sycl/source/detail/config.def:42]: Good point, done now.

File: sycl/source/detail/config.hpp (hunk 4 above)
Comment (sommerlukas) [reply sycl/source/detail/config.hpp]: Agreed, error is caught now.

File: sycl/source/detail/config.hpp [hunk 5]
```diff
@@ -579,6 +579,43 @@ template <> class SYCLConfig<SYCL_REDUCTION_PREFERRED_WORKGROUP_SIZE> {
...
+    int Result = DefaultValue;
+    try {
+      Result = std::stoi(ValStr);
+    } catch (...) {
+      throw invalid_parameter_error(
+          "Invalid value for SYCL_ENABLE_FUSION_CACHING environment variable: "
+          "value should be '0' or '1'",
+          PI_ERROR_INVALID_VALUE);
+    }
+    return Result;
```
Comment (steffenlarsen) [inline sycl/source/detail/config.hpp:603]: Nit; technically we are allowing any integer value here, so should it check that it's one of the valid ones or maybe just say "Disable ('0') or enable (any non-zero integer value) ..." in the docs?

**Review by steffenlarsen — APPROVED** [review]
Changes look good to me, apart from a small nit.

File: sycl/source/detail/config.hpp (hunk 5 above)
Comment (sommerlukas) [reply sycl/source/detail/config.hpp]: That's true, I adopted what is done for `SYCL_ENABLE_DEFAULT_CONTEXTS` to allow only `1` and `0` values.

**Review by steffenlarsen — APPROVED** [review]

**Comment by sommerlukas** [comment]
@intel/llvm-gatekeepers Can someone please merge, I'm not authorized.

**Pull request merged**